File: android/src/main/java/expo/modules/crispsdk/CompanyParser.kt
package expo.modules.crispsdk

import im.crisp.client.external.data.Company
import im.crisp.client.external.data.Employment
import im.crisp.client.external.data.Geolocation
import java.net.URL

/**
 * Parser for Company objects from JavaScript
 * Converts Map<String, Any?> to Crisp SDK Company class
 *
 * TypeScript interface:
 * interface Company {
 *   name: string;
 *   url?: string;
 *   companyDescription?: string;
 *   employment?: Employment;
 *   geolocation?: Geolocation;
 * }
 */
object CompanyParser {
    fun fromMap(map: Map<String, Any?>): Company {
        val name = map["name"] as? String ?: ""
        val url = (map["url"] as? String)?.let {
            try { URL(it) } catch (e: Exception) { null }
        }
        val companyDescription = map["companyDescription"] as? String

        val employment = (map["employment"] as? Map<String, Any?>)?.let {
            EmploymentParser.fromMap(it)
        }

        val geolocation = (map["geolocation"] as? Map<String, Any?>)?.let {
            GeolocationParser.fromMap(it)
        }

        return Company(name, url, companyDescription, employment, geolocation)
    }
}

/**
 * Parser for Employment objects from JavaScript
 * Converts Map<String, Any?> to Crisp SDK Employment class
 *
 * TypeScript interface:
 * interface Employment {
 *   title?: string;
 *   role?: string;
 * }
 */
object EmploymentParser {
    fun fromMap(map: Map<String, Any?>): Employment {
        val title = map["title"] as? String
        val role = map["role"] as? String
        return Employment(title, role)
    }
}

/**
 * Parser for Geolocation objects from JavaScript
 * Converts Map<String, Any?> to Crisp SDK Geolocation class
 *
 * TypeScript interface:
 * interface Geolocation {
 *   country?: string;
 *   city?: string;
 * }
 */
object GeolocationParser {
    fun fromMap(map: Map<String, Any?>): Geolocation {
        val country = map["country"] as? String
        val city = map["city"] as? String
        return Geolocation(country, city)
    }
}

File: ios/CompanyParser.swift
import Crisp
import Foundation

enum CompanyParser {
  static func fromDictionary(_ dict: [String: Any]) -> Company {
    let name = dict["name"] as? String

    var url: URL?
    if let urlString = dict["url"] as? String, !urlString.isEmpty {
      url = URL(string: urlString)
    }

    let companyDescription = dict["companyDescription"] as? String

    var employment: Employment?
    if let employmentDict = dict["employment"] as? [String: Any] {
      let title = employmentDict["title"] as? String
      let role = employmentDict["role"] as? String
      employment = Employment(title: title, role: role)
    }

    var geolocation: Geolocation?
    if let geoDict = dict["geolocation"] as? [String: Any] {
      let city = geoDict["city"] as? String
      let country = geoDict["country"] as? String
      geolocation = Geolocation(city: city, country: country)
    }

    return Company(
      name: name,
      url: url,
      companyDescription: companyDescription,
      employment: employment,
      geolocation: geolocation
    )
  }
}
